V compile fails with option type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.10 dd859ea, press to see full `v doctor` output</summary>

|V full version      |V 0.4.10 dd859eae55cf4e69346851fe285b9af104c8ffb7.dd859ea
|:-------------------|:-------------------
|OS                  |linux, Linux Mint 22
|Processor           |16 cpus, 64bit, little endian, AMD Ryzen 7 5800H with Radeon Graphics
|Memory              |2.34GB/13.5GB
|                    |
|V executable        |/home/joao/dev/Morpheus/sdks/v/v
|V last modified time|2025-05-06 21:52:42
|                    |
|V home dir          |OK, value: /home/joao/dev/Morpheus/sdks/v
|VMODULES            |OK, value: /home/joao/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/joao/dev/Morpheus/examples/slimcc-lua
|                    |
|Git version         |git version 2.43.0
|V git status        |acdce4e1-dirty
|.git/config present |false
|                    |
|cc version          |cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0
|gcc version         |gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0
|clang version       |Ubuntu clang version 18.1.3 (1ubuntu1)
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (Ubuntu GLIBC 2.39-0ubuntu8.3) 2.39

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  syntax_tree.v && /home/joao/dev/Morpheus/examples/slimcc-lua/syntax_tree`
```v
// Simple arena allocator
struct ArenaChunk {
mut:
    next   ?&ArenaChunk
    used   int
    cap    int
    data   byteptr
}

struct Arena {
mut:
    head ?&ArenaChunk
}

fn arena_init(mut arena Arena, first_capacity int) {
    chunk := &ArenaChunk{
        next: 0
        used: 0
        cap: first_capacity
        data: unsafe { malloc(first_capacity) }
    }
    arena.head = chunk
}

fn main() {
    mut green_arena := Arena{}
    arena_init(mut green_arena, 64 * 1024)
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/syntax_tree.01JVDY2H3VRH58FYNAC987V2PJ.tmp.c:5644: error: lvalue expected
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

```c
VV_LOCAL_SYMBOL void main__arena_init(main__Arena* arena, int first_capacity) {
	_option_main__ArenaChunk_ptr _t1;
	_option_ok(&(main__ArenaChunk*[]) { &0 }, (_option*)(&_t1), sizeof(main__ArenaChunk*));
	main__ArenaChunk* chunk = ((main__ArenaChunk*)memdup(&(main__ArenaChunk){.next = _t1,.used = 0,.cap = first_capacity,.data = _v_malloc(first_capacity),}, sizeof(main__ArenaChunk)));
	_option_main__ArenaChunk_ptr _t2;
	_option_ok(&(main__ArenaChunk*[]) { chunk }, (_option*)(&_t2), sizeof(main__ArenaChunk*));
	arena->head = _t2;
}
```

**What did you expect to see?**
V error
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2333,7 +2333,8 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 						}
 					}
 				}
-				if expr_typ != ast.nil_type && ret_typ.nr_muls() > expr_typ.nr_muls() {
+				if !expr.is_literal() && expr_typ != ast.nil_type
+					&& ret_typ.nr_muls() > expr_typ.nr_muls() {
 					g.write('&'.repeat(ret_typ.nr_muls() - expr_typ.nr_muls()))
 				}
 			}

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2333,7 +2333,7 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 						}
 					}
 				}
-				if ret_typ.nr_muls() > expr_typ.nr_muls() {
+				if expr_typ != ast.nil_type && ret_typ.nr_muls() > expr_typ.nr_muls() {
 					g.write('&'.repeat(ret_typ.nr_muls() - expr_typ.nr_muls()))
 				}
 			}

--- a/vlib/v/tests/options/option_ptr_nil_test.v
+++ b/vlib/v/tests/options/option_ptr_nil_test.v
@@ -1,0 +1,21 @@
+struct Foo {
+	name ?&string
+}
+
+struct Foo2 {
+mut:
+	name ?&string
+}
+
+fn test_1() {
+	_ := Foo{
+		name: unsafe { nil }
+	}
+}
+
+fn test_2() {
+	mut foo := Foo2{}
+	unsafe {
+		foo.name = nil
+	}
+}

--- a/vlib/v/tests/options/option_ptr_zero_test.v
+++ b/vlib/v/tests/options/option_ptr_zero_test.v
@@ -1,0 +1,28 @@
+// Simple arena allocator
+struct ArenaChunk {
+mut:
+	next ?&ArenaChunk
+	used int
+	cap  int
+	data byteptr
+}
+
+struct Arena {
+mut:
+	head ?&ArenaChunk
+}
+
+fn arena_init(mut arena Arena, first_capacity int) {
+	chunk := &ArenaChunk{
+		next: 0
+		used: 0
+		cap:  first_capacity
+		data: unsafe { malloc(first_capacity) }
+	}
+	arena.head = chunk
+}
+
+fn test_main() {
+	mut green_arena := Arena{}
+	arena_init(mut green_arena, 64 * 1024)
+}

